Assets/Documentation/man/man3/GameManager.3
.TH "GameManager" 3 "Version 1.0.0" "KiBoard GDD & Technical Documentation" \" -*- nroff -*-
.ad l
.nh
.SH NAME
GameManager
.SH SYNOPSIS
.br
.PP
.PP
Inherits MonoBehaviour\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "void \fBAssignFpsText\fP (TMP_InputField text)"
.br
.RI "Assigns text parameter's \&.text feild to the current target FPS\&. "
.ti -1c
.RI "void \fBSetFPS\fP (string str)"
.br
.RI "Sets Application\&.targetFrameRate to 'str' parsed\&. "
.ti -1c
.RI "void \fBSetInputDelay\fP (string delayStr)"
.br
.RI "Sets PlayerSettings\&.inputDelay to 'delayStr' parsed\&. "
.ti -1c
.RI "void \fBUpdateSongTime\fP ()"
.br
.RI "Updates \fBsongTime\fP variable with Time\&.deltaTime\&. "
.ti -1c
.RI "void \fBSetSongTotalNotes\fP (int noteCount)"
.br
.RI "Sets \fBtotalNotes\fP to the total number of notes in the song\&. "
.ti -1c
.RI "IEnumerator \fBPrepareNotes\fP (float BPM, List< \fBNoteEventInfo\fP > noteEvents, bool isPreview)"
.br
.RI "Prepares notes for a song to be played\&. "
.ti -1c
.RI "void \fBStopReadiedNotes\fP ()"
.br
.RI "Stops all coroutines for preparing notes\&. "
.ti -1c
.RI "void \fBUpdatePlayerScore\fP (string score)"
.br
.RI "Updates the player's score based on the hit timing\&. "
.ti -1c
.RI "string \fBConvertNoteNumberToName\fP (int noteNumber)"
.br
.RI "Converts a MIDI note number to its corresponding name\&. "
.ti -1c
.RI "void \fBOnSongEnd\fP ()"
.br
.RI "Coroutine for handling song end actions\&. "
.ti -1c
.RI "void \fBReturnToSongSelection\fP ()"
.br
.RI "Returns to the song selection scene\&. "
.ti -1c
.RI "void \fBStopSong\fP ()"
.br
.RI "Stops all prepared notes and stops updating of songTime\&. "
.ti -1c
.RI "void \fBRefreshJsonFiles\fP ()"
.br
.RI "Reloads all json file versions of the currently selected song\&. "
.ti -1c
.RI "void \fBSetBeatsBeforeDrop\fP (string num)"
.br
.RI "Sets \fBbeatsBeforeNote\fP\&. "
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static bool \fBCheckSpawnNote\fP (\fBNoteEventInfo\fP note)"
.br
.RI "Used to check if a noteEvent is storing Tempo Data or Note Data\&. i assign tempo changes by setting the \fBNoteEventInfo\&.startTime\fP to negative infinity and the \fBNoteEventInfo\&.noteNumber\fP to min value, with the tempo as the end time\&. "
.ti -1c
.RI "static int \fBConvertNoteNameToNumber\fP (string noteName)"
.br
.RI "Converts a note name to its corresponding MIDI note number\&. "
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "GameObject \fBnotePrefab\fP"
.br
.RI "prefab used for instantiating notes\&. "
.ti -1c
.RI "\fBSongScore\fP \fBcurrentSongScore\fP"
.br
.RI "\fBSongScore\fP for currently playing song\&. "
.ti -1c
.RI "\fBSongScore\fP \fBselectedSongHighScore\fP"
.br
.RI "\fBSongScore\fP storing the High score for the currently selected song\&. "
.ti -1c
.RI "\fBCombo\fP \fBcombo\fP = new()"
.br
.RI "Current \fBCombo\fP for the song\&. "
.ti -1c
.RI "bool \fBinEditor\fP"
.br
.RI "Is the game in the SongEditor State? "
.ti -1c
.RI "bool \fBstartTimer\fP"
.br
.RI "Should \fBsongTime\fP be updated? "
.ti -1c
.RI "bool \fBisCurSongPreview\fP"
.br
.RI "Is the currently playing song appearing in the preview window in the Song Selection scene\&. "
.ti -1c
.RI "float \fBsongTime\fP"
.br
.RI "Elapsed duration (seconds) into the currently playing song\&. "
.ti -1c
.RI "float \fBmodifiedNoteScale\fP"
.br
.RI "Note scaling factor calculated by the song BPM at the time of the notes appearence in the song\&.
.br
 See also 
.PP
\fBSee also\fP
.RS 4
\fBbaseNoteScalingFactor\fP
.PP
.RE
.PP
"
.ti -1c
.RI "int \fBbeatsBeforeNote\fP = 4"
.br
.RI "Number of beats a note should spawn before being visible on screen\&. 
.br
 Modification changes the speed notes fall, and changes the amount of time players have to react to notes\&. "
.ti -1c
.RI "int \fBtotalNotes\fP"
.br
.RI "Total number of notes in the currently playing song\&. "
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static \fBGameManager\fP \fBinstance\fP"
.br
.RI "singleton reference\&. "
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBAwake\fP ()"
.br
.ti -1c
.RI "void \fBUpdate\fP ()"
.br
.ti -1c
.RI "void \fBAssignToPreviewLayer\fP (GameObject obj)"
.br
.RI "Assigns 'obk' to the layer 'PreviewLayer', then moves it and all children objects to the song preview scene\&. "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "float \fBscreenHeight\fP"
.br
.RI "Height of the viewport in Units, as viewed from an angle in the game view\&. "
.ti -1c
.RI "float \fBfallSpeed\fP"
.br
.ti -1c
.RI "float \fBspawnOffset\fP = 2f"
.br
.ti -1c
.RI "float \fBbaseNoteScalingFactor\fP = 5\&.4f"
.br
.RI "Base scaling factor for notes to be modified by BPM of the song\&.
.br
 do not ask me where the 5\&.4 came from i do not know\&. "
.ti -1c
.RI "Transform \fBnoteHolder\fP"
.br
.ti -1c
.RI "List< Coroutine > \fBreadiedNotes\fP = new()"
.br
.RI "List holding all note prepared but not yet spawned\&. "
.ti -1c
.RI "string[] \fBnoteNames\fP = { 'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B', }"
.br
.RI "String array containing the names of notes accoring to their indecies, 0 being C\&. "
.in -1c
.SS "Static Private Attributes"

.in +1c
.ti -1c
.RI "static Dictionary< string, int > \fBnameToNoteMap\fP"
.br
.RI "Dictionary mapping of note names to note numbers\&. "
.in -1c
.SH "Member Function Documentation"
.PP 
.SS "void GameManager\&.AssignFpsText (TMP_InputField text)"

.PP
Assigns text parameter's \&.text feild to the current target FPS\&. 
.PP
\fBParameters\fP
.RS 4
\fItext\fP TextMeshPro to have string assigned to\&.
.RE
.PP

.nf
124     {
125         text\&.text = Application\&.targetFrameRate\&.ToString();
126     }
.PP
.fi

.SS "void GameManager\&.AssignToPreviewLayer (GameObject obj)\fR [private]\fP"

.PP
Assigns 'obk' to the layer 'PreviewLayer', then moves it and all children objects to the song preview scene\&. 
.PP
\fBParameters\fP
.RS 4
\fIobj\fP object to have layer assigned\&.
.RE
.PP

.nf
322     {
323 
324         // Assign the object to the "PreviewLayer"
325         LayerMask layer = LayerMask\&.NameToLayer("PreviewLayer");
326         try { SceneManager\&.MoveGameObjectToScene(obj, MidiInput\&.instance\&.currentPreview); } catch { }
327 
328         foreach (Transform child in obj\&.GetComponentsInChildren<Transform>(true))
329         {
330             child\&.gameObject\&.layer = layer;
331         }
332     }
.PP
.fi

.PP
Referenced by \fBPrepareNotes()\fP\&.
.SS "void GameManager\&.Awake ()\fR [private]\fP"

.nf
112     {
113         if (instance == null)
114         {
115             instance = this;
116         }
117         else { Destroy(gameObject); }
118     }
.PP
.fi

.SS "static bool GameManager\&.CheckSpawnNote (\fBNoteEventInfo\fP note)\fR [static]\fP"

.PP
Used to check if a noteEvent is storing Tempo Data or Note Data\&. i assign tempo changes by setting the \fBNoteEventInfo\&.startTime\fP to negative infinity and the \fBNoteEventInfo\&.noteNumber\fP to min value, with the tempo as the end time\&. 
.PP
\fBParameters\fP
.RS 4
\fInote\fP note to check for tempo data\&.
.RE
.PP
\fBReturns\fP
.RS 4
True = note data, false = tempo data\&.
.RE
.PP

.nf
184     {
185         return (note\&.startTime == float\&.NegativeInfinity && note\&.noteNumber == int\&.MinValue);
186     }
.PP
.fi

.PP
Referenced by \fBPrepareNotes()\fP\&.
.SS "static int GameManager\&.ConvertNoteNameToNumber (string noteName)\fR [static]\fP"

.PP
Converts a note name to its corresponding MIDI note number\&. 
.PP
\fBParameters\fP
.RS 4
\fInoteName\fP The note name (e\&.g\&., C, C#, D)\&.
.RE
.PP
\fBReturns\fP
.RS 4
The MIDI note number\&.
.RE
.PP

.nf
381     {
382         if (!char\&.IsDigit(noteName[0]))
383         {
384             int octave = 0;
385 
386             for (int i = 0; i < noteName\&.Length; i++)
387             {
388                 if (char\&.IsDigit(noteName[i]))
389                 {
390                     octave = noteName[i];
391                 }
392             }
393             nameToNoteMap\&.TryGetValue(noteName\&.ToLower(), out int value);
394 
395             return value + (12 * octave);
396         }
397         return int\&.Parse(noteName);
398 
399     }
.PP
.fi

.SS "string GameManager\&.ConvertNoteNumberToName (int noteNumber)"

.PP
Converts a MIDI note number to its corresponding name\&. 
.PP
\fBParameters\fP
.RS 4
\fInoteNumber\fP The MIDI note number\&.
.RE
.PP
\fBReturns\fP
.RS 4
The note name (e\&.g\&., C, C#, D)\&.
.RE
.PP

.nf
363     {
364 
365         int octave = (noteNumber / 12) \- 1;
366 
367         int noteIndex = (noteNumber) % 12;
368         string noteName = noteNames[noteIndex];
369         if (GameSettings\&.usePiano)
370         {
371             return $"{noteName} {octave}";
372         }
373         return $"{noteName}";
374     }
.PP
.fi

.SS "void GameManager\&.OnSongEnd ()"

.PP
Coroutine for handling song end actions\&. 
.nf
405     {
406         startTimer = false;
407         MidiInput\&.instance\&.takeInput = false;
408         int[] score = currentSongScore\&.GetScoreArray(totalNotes);
409 
410         if (!isCurSongPreview) { EndSongMessage\&.instance?\&.ShowScore($"Total Score: {score[0]}\\nPerfect: {score[1]}\\nGood: {score[2]}\\nOkay: {score[3]}\\nExtra: {score[4]}\\nMissed: {score[5]}\\nLongest Combo: {combo\&.highestCount}", currentSongScore\&.FinalizeScore()); }
411         else { FindObjectOfType<EndPreview>()?\&.EndPreviewFade(); }
412 
413         if (!Replay\&.isPlayingReplay) { MidiDataHandler\&.SaveNoteEventData("\&.replay", Replay\&.instance\&.replayNoteData); } // only record replays if you arent playing back a replay\&.
414 
415     }
.PP
.fi

.SS "IEnumerator GameManager\&.PrepareNotes (float BPM, List< \fBNoteEventInfo\fP > noteEvents, bool isPreview)"

.PP
Prepares notes for a song to be played\&. 
.PP
\fBParameters\fP
.RS 4
\fIBPM\fP Starting BPM of selected song
.br
\fInoteEvents\fP List of \fBNoteEventInfo\fP's to be played\&.
.br
\fIisPreview\fP Is the song in the Song Selection Preview Window?
.RE
.PP
\fBReturns\fP
.RS 4
.RE
.PP

.nf
195     {
196         if (noteEvents == null) { Debug\&.Log("gameloop noteEvents null"); yield break; }
197         GameType? gameType = GameSettings\&.gameType = GameSettings\&.usePiano ? GameType\&.Key88 : GameType\&.Key12;
198 
199         SetSongTotalNotes(noteEvents\&.Count);
200         SongScore songScore = new();
201         if (isPreview) { isCurSongPreview = true; } else { isCurSongPreview = false; }
202         songTime = \-3f \- (130 / BPM);
203         if (!isCurSongPreview) { Replay\&.recordReplay = true; Replay\&.StartReplayCapture(); }
204         else { Replay\&.recordReplay = false; }
205         screenHeight = 40\&.16f;//2f * Camera\&.main\&.orthographicSize;
206         AssignSongValues(BPM);
207 
208         yield return new WaitForSecondsRealtime(1f);
209         yield return new WaitUntil(() => (Input\&.anyKeyDown || MidiInput\&.instance\&.GetAnyNoteActive()) || isPreview);
210         startTimer = true;
211         StopReadiedNotes();
212         if (gameType == GameType\&.Key88) { noteEvents\&.ForEach(noteEvent => readiedNotes\&.Add(StartCoroutine(ReadyNote88(noteEvent\&.startTime, noteEvent)))); }
213         else if (gameType == GameType\&.Key12) { noteEvents\&.ForEach(noteEvent => readiedNotes\&.Add(StartCoroutine(ReadyNote12(noteEvent\&.startTime, noteEvent)))); }
214         else
215         {
216             Debug\&.LogError("Game type not 88 or 12 key mode\&.");
217         }
218 
219         // Game loop is finished
220         yield return null;
221 
222         void AssignSongValues(float BPM)
223         {
224 
225             spawnOffset = (beatsBeforeNote * 60f / BPM);
226             // Calculate the speed based on the distance and duration
227             fallSpeed = (screenHeight / spawnOffset);
228 
229             modifiedNoteScale = baseNoteScalingFactor * (130 / BPM);
230         }
231 
232         float TempoChange(NoteEventInfo note)
233         {
234             BPM = note\&.endTime;
235             float so = (beatsBeforeNote * 60f / BPM);
236             fallSpeed = (screenHeight / so);
237             modifiedNoteScale = baseNoteScalingFactor * (130 / BPM);
238             return so;
239         }
240         IEnumerator ReadyNote88(float spawnTime, NoteEventInfo noteEvent)
241         {
242 
243             if (CheckSpawnNote(noteEvent))
244             {
245                 spawnOffset = TempoChange(noteEvent);
246                 yield break;
247 
248             }
249             float trueSpawnTime = spawnTime \- spawnOffset;
250             yield return new WaitUntil(() => songTime >= trueSpawnTime);
251             SpawnNote88(noteEvent);
252         }
253         IEnumerator ReadyNote12(float spawnTime, NoteEventInfo noteEvent)
254         {
255             if (CheckSpawnNote(noteEvent))
256             {
257                 spawnOffset = TempoChange(noteEvent);
258                 AssignSongValues(noteEvent\&.endTime);
259 
260             }
261             float trueSpawnTime = spawnTime \- spawnOffset;
262             yield return new WaitUntil(() => songTime >= trueSpawnTime);
263             SpawnNote12(noteEvent);
264         }
265 
266         void SpawnNote88(NoteEventInfo noteEvent)
267         {
268 
269             // scale/length of the note deterimned by the note duration, and a scaling factor (~~~~~~~~~~~~~~~~~BASE THIS ON MF BPM)``````````````````````````````````````````````````````````````````````````````````
270             float noteScale = (noteEvent\&.endTime \- noteEvent\&.startTime) * modifiedNoteScale;
271             //spawn a note and store a reference\&.
272             GameObject noteInstance = Instantiate(notePrefab, new Vector3(\-13\&.2f + (0\&.20505f * (noteEvent\&.noteNumber)), (screenHeight) + (noteScale / 2) \- 2\&.5f, 0f), Quaternion\&.identity);
273             FallingNote fallingNote = noteInstance\&.GetComponent<FallingNote>();
274             SpriteRenderer spriteRenderer = noteInstance\&.GetComponent<SpriteRenderer>();
275             fallingNote\&.velocity = fallSpeed; // set falling speed of the note to the value calculated in AssignSongValues()
276             fallingNote\&.maxYBound = spriteRenderer\&.bounds\&.max\&.y; //Used to determine when a note is far enough off screen to be destroyed\&.
277             fallingNote\&.GetComponentInChildren<NoteShadow>()\&.SetShadowSize(noteScale + 0\&.075f);
278             if (noteHolder != null)
279             {
280                 noteInstance\&.transform\&.SetParent(noteHolder\&.transform, false);
281             }
282             spriteRenderer\&.size = new Vector2(spriteRenderer\&.size\&.x, noteScale);
283 
284 
285             if (noteEvent == noteEvents[noteEvents\&.Count \- 1]) // check if note is the final note\&.
286             {
287                 fallingNote\&.isLast = true;// set flag to end song after the last note is destroyed\&.
288 
289             }
290             if (isPreview) AssignToPreviewLayer(noteInstance);
291         }

Assets/Scripts/GameManager.cs
using NAudio.Midi;
using System.Collections;
using System.Collections.Generic;
using TMPro;
using UnityEngine;
using UnityEngine.SceneManagement;

public class GameManager : MonoBehaviour
{
    /// <summary>
    /// singleton reference.
    /// </summary>
    public static GameManager instance;
    /// <summary>
    /// keyPrefab used for instantiating notes.
    /// </summary>
    public GameObject notePrefab;
    /// <summary>
    /// <see cref="SongScore"/> for currently playing song.
    /// </summary>
    public SongScore currentSongScore;
    /// <summary>
    /// <see cref="SongScore"/> storing the High score for the currently selected song.
    /// </summary>
    public SongScore selectedSongHighScore;
    /// <summary>
    /// Current <see cref="Combo"/> for the song.
    /// </summary>
    public Combo combo = new();
    /// <summary>
    /// Is the game in the SongEditor State?
    /// </summary>
    public bool inEditor;
    /// <summary>
    /// Should <see cref="songTime"/> be updated?
    /// </summary>
    public bool startTimer;
    /// <summary>
    /// Is the currently playing song appearing in the preview window in the Song Selection scene.
    /// </summary>
    public bool isCurSongPreview;
    /// <summary>
    /// Elapsed duration (seconds) into the currently playing song.
    /// </summary>
    public float songTime;
    /// <summary>
    /// Note scaling factor calculated by the song BPM at the time of the notes appearence in the song.<br/>
    /// See also <seealso cref="baseNoteScalingFactor"/>
    /// </summary>
    public float modifiedNoteScale;
    /// <summary>
    /// Number of beats a note should spawn before being visible on screen. <br/>
    /// Modification changes the speed notes fall, and changes the amount of time players have to react to notes.
    /// </summary>
    public int beatsBeforeNote = 4;
    /// <summary>
    /// Total number of notes in the currently playing song.
    /// </summary>
    public int totalNotes;

    /// <summary>
    /// Height of the viewport in Units, as viewed from an angle in the game view.
    /// </summary>
    float screenHeight;

    float fallSpeed;
    float spawnOffset = 2f;
    /// <summary>
    /// Base scaling factor for notes to be modified by BPM of the song.<br/>
    /// do not ask me where the 5.4 came from i do not know.
    /// </summary>
    float baseNoteScalingFactor = 5.4f;
    [SerializeField] Transform noteHolder;
    /// <summary>
    /// List holding all note prepared but not yet spawned.
    /// </summary>
    List<Coroutine> readiedNotes = new();
    /// <summary>
    /// String array containing the names of notes accoring to their indecies, 0 being C.
    /// </summary>
    string[] noteNames = { "C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B", };

    /// <summary>
    /// Dictionary mapping of note names to note numbers.
    /// </summary>
    static Dictionary<string, int> nameToNoteMap = new()
    {
        { "cb", 12 },
        { "c", 1 },
        { "c#", 2 },
        { "db", 2 },
        { "d", 3 },
        { "d#", 4 },
        { "eb", 4 },
        { "e", 5 },
        { "e#", 6 },
        { "fb", 5 },
        { "f", 6 },
        { "f#", 7 },
        { "gb", 7 },
        { "g", 8 },
        { "g#", 9 },
        { "ab", 9 },
        { "a", 10 },
        { "a#", 11 },
        { "bb", 11 },
        { "b", 12 },
        { "b#", 1 },
    };

    private void Awake()
    {
        if (instance == null)
        {
            instance = this;
        }
        else { Destroy(gameObject); }
    }
    /// <summary>
    /// Assigns text parameter's .text feild to the current target FPS.
    /// </summary>
    /// <param name="text">TextMeshPro to have string assigned to.</param>
    public void AssignFpsText(TMP_InputField text)
    {
        text.text = Application.targetFrameRate.ToString();
    }
    /// <summary>
    /// Sets <see cref="Application.targetFrameRate"/> to "str" parsed.
    /// </summary>
    /// <param name="str">Value to set FPS to.</param>
    public void SetFPS(string str)
    {
        if (int.TryParse(str, out int newFps))
        {
            Application.targetFrameRate = newFps;
        }
        else { Debug.LogError($"FPS Cap of : {str} is invalid"); }
    }
    /// <summary>
    /// Sets <see cref="PlayerSettings.inputDelay"/> to "delayStr" parsed.
    /// </summary>
    /// <param name="delayStr"></param>
    public void SetInputDelay(string delayStr)
    {
        if (int.TryParse(delayStr, out int delay))
        {
            PlayerSettings.inputDelay = delay;
        }
        else { Debug.Log("Error Parsing input delay"); }
    }

    private void Update()
    {
        UpdateSongTime();
    }
    /// <summary>
    /// Updates <see cref="songTime"/> variable with Time.deltaTime.
    /// </summary>
    public void UpdateSongTime()
    {
        if (startTimer)
        {
            songTime += Time.deltaTime;
        }
    }
    /// <summary>
    /// Sets <see cref="totalNotes"/> to the total number of notes in the song.
    /// </summary>
    /// <param name="noteCount">The total number of notes.</param>

    public void SetSongTotalNotes(int noteCount)
    {
        totalNotes = noteCount;
    }

    /// <summary>
    /// Used to check if a noteEvent is storing Tempo Data or Note Data.
    /// i assign tempo changes by setting the <see cref="NoteEventInfo.startTime"/> to negative infinity and the <see cref="NoteEventInfo.noteNumber"/> to min value, with the tempo as the end time.
    /// 
    /// </summary>
    /// <param name="note">note to check for tempo data.</param>
    /// <returns>True = note data, false = tempo data.</returns>
    public static bool CheckSpawnNote(NoteEventInfo note)
    {
        return (note.startTime == float.NegativeInfinity && note.noteNumber == int.MinValue);
    }
    /// <summary>
    /// Prepares notes for a song to be played.
    /// </summary>
    /// <param name="BPM">Starting BPM of selected song</param>
    /// <param name="noteEvents">List of <see cref="NoteEventInfo"/>'s to be played.</param>
    /// <param name="isPreview">Is the song in the Song Selection Preview Window?</param>
    /// <returns></returns>
    public IEnumerator PrepareNotes(float BPM, List<NoteEventInfo> noteEvents, bool isPreview) // TEMP 0.5f, change to 5.4f i think`````````````````````````````````````````````````````````````````````````````````````````````````````````
    {
        if (noteEvents == null) { Debug.Log("gameloop noteEvents null"); yield break; }
        GameType? gameType = GameSettings.gameType = GameSettings.usePiano ? GameType.Key88 : GameType.Key12;

        SetSongTotalNotes(noteEvents.Count);
        SongScore songScore = new();
        if (isPreview) { isCurSongPreview = true; } else { isCurSongPreview = false; }
        songTime = -3f - (130 / BPM);
        if (!isCurSongPreview) { Replay.recordReplay = true; Replay.StartReplayCapture(); }
        else { Replay.recordReplay = false; }
        screenHeight = 40.16f;//2f * Camera.main.orthographicSize;
        AssignSongValues(BPM);

        yield return new WaitForSecondsRealtime(1f);
        yield return new WaitUntil(() => (Input.anyKeyDown || MidiInput.instance.GetAnyNoteActive()) || isPreview);
        startTimer = true;
        StopReadiedNotes();
        if (gameType == GameType.Key88) { noteEvents.ForEach(noteEvent => readiedNotes.Add(StartCoroutine(ReadyNote88(noteEvent.startTime, noteEvent)))); }
        else if (gameType == GameType.Key12) { noteEvents.ForEach(noteEvent => readiedNotes.Add(StartCoroutine(ReadyNote12(noteEvent.startTime, noteEvent)))); }
        else
        {
            Debug.LogError("Game type not 88 or 12 key mode.");
        }

        // Game loop is finished
        yield return null;

        void AssignSongValues(float BPM)
        {

            spawnOffset = (beatsBeforeNote * 60f / BPM);
            // Calculate the speed based on the distance and duration
            fallSpeed = (screenHeight / spawnOffset);

            modifiedNoteScale = baseNoteScalingFactor * (130 / BPM);
        }

        float TempoChange(NoteEventInfo note)
        {
            BPM = note.endTime;
            float so = (beatsBeforeNote * 60f / BPM);
            fallSpeed = (screenHeight / so);
            modifiedNoteScale = baseNoteScalingFactor * (130 / BPM);
            return so;
        }
        IEnumerator ReadyNote88(float spawnTime, NoteEventInfo noteEvent)
        {

            if (CheckSpawnNote(noteEvent))
            {
                spawnOffset = TempoChange(noteEvent);
                yield break;

            }
            float trueSpawnTime = spawnTime - spawnOffset;
            yield return new WaitUntil(() => songTime >= trueSpawnTime);
            SpawnNote88(noteEvent);
        }
        IEnumerator ReadyNote12(float spawnTime, NoteEventInfo noteEvent)
        {
            if (CheckSpawnNote(noteEvent))
            {
                spawnOffset = TempoChange(noteEvent);
                AssignSongValues(noteEvent.endTime);

            }
            float trueSpawnTime = spawnTime - spawnOffset;
            yield return new WaitUntil(() => songTime >= trueSpawnTime);
            SpawnNote12(noteEvent);
        }

        void SpawnNote88(NoteEventInfo noteEvent)
        {

            // scale/length of the note deterimned by the note duration, and a scaling factor (~~~~~~~~~~~~~~~~~BASE THIS ON MF BPM)``````````````````````````````````````````````````````````````````````````````````
            float noteScale = (noteEvent.endTime - noteEvent.startTime) * modifiedNoteScale;
            //spawn a note and store a reference.
            GameObject noteInstance = Instantiate(notePrefab, new Vector3(-13.2f + (0.20505f * (noteEvent.noteNumber)), (screenHeight) + (noteScale / 2) - 2.5f, 0f), Quaternion.identity);
            FallingNote fallingNote = noteInstance.GetComponent<FallingNote>();
            SpriteRenderer spriteRenderer = noteInstance.GetComponent<SpriteRenderer>();
            fallingNote.velocity = fallSpeed; // set falling speed of the note to the value calculated in AssignSongValues()
            fallingNote.maxYBound = spriteRenderer.bounds.max.y; //Used to determine when a note is far enough off screen to be destroyed.
            fallingNote.GetComponentInChildren<NoteShadow>().SetShadowSize(noteScale + 0.075f);
            if (noteHolder != null)
            {
                noteInstance.transform.SetParent(noteHolder.transform, false);
            }
            spriteRenderer.size = new Vector2(spriteRenderer.size.x, noteScale);


            if (noteEvent == noteEvents[noteEvents.Count - 1]) // check if note is the final note.
            {
                fallingNote.isLast = true;// set flag to end song after the last note is destroyed.

            }
            if (isPreview) AssignToPreviewLayer(noteInstance);
        }
        void SpawnNote12(NoteEventInfo noteEvent)
        {
            // scale/length of the note deterimned by the note duration, and a scaling factor  (~~~~~~~~~~~~~~~~~BASE THIS ON MF BPM)``````````````````````````````````````````````````````````````````````````````````
            float noteScale = (noteEvent.endTime - noteEvent.startTime) * modifiedNoteScale;
            //spawn a note and store a reference.
            GameObject noteInstance = Instantiate(notePrefab, new Vector3(-5.6f + (1 * (noteEvent.noteNumber % 12)), (screenHeight) + (noteScale / 2) - 2.5f, 0f), Quaternion.identity);
            FallingNote fallingNote = noteInstance.GetComponent<FallingNote>();
            SpriteRenderer spriteRenderer = noteInstance.GetComponent<SpriteRenderer>();
            fallingNote.velocity = fallSpeed; // set falling speed of the note to the value calculated in AssignSongValues()
            fallingNote.maxYBound = spriteRenderer.bounds.max.y; //Used to determine when a note is far enough off screen to be destroyed.
            fallingNote.GetComponentInChildren<NoteShadow>().SetShadowSize(noteScale + 0.075f);

            spriteRenderer.size = new Vector2(4.5f, noteScale);


            if (noteEvent == noteEvents[noteEvents.Count - 1]) // check if note is the final note.
            {
                fallingNote.isLast = true;// set flag to end song after the last note is destroyed.


            }
            if (isPreview) AssignToPreviewLayer(noteInstance);
        }
    }

    /// <summary>
    /// Assigns "obk" to the layer "PreviewLayer", then moves it and all children objects to the song preview scene.
    /// </summary>
    /// <param name="obj">object to have layer assigned.</param>
    void AssignToPreviewLayer(GameObject obj)
    {

        // Assign the object to the "PreviewLayer"
        LayerMask layer = LayerMask.NameToLayer("PreviewLayer");
        try { SceneManager.MoveGameObjectToScene(obj, MidiInput.instance.currentPreview); } catch { }

        foreach (Transform child in obj.GetComponentsInChildren<Transform>(true))
        {
            child.gameObject.layer = layer;
        }
    }
    /// <summary>
    /// Stops all coroutines for preparing notes.
    /// </summary>
    public void StopReadiedNotes()
    {
        foreach (Coroutine c in readiedNotes)
        {
            if (c != null)
            {
                StopCoroutine(c);
            }
        }
        readiedNotes.Clear();
    }
    /// <summary>
    /// Updates the player's score based on the hit timing.
    /// </summary>
    /// <param name="score">The hit score (Perfect, Good, Okay, Miss).</param>
    public void UpdatePlayerScore(string score)
    {
        currentSongScore ??= new();
        currentSongScore.AddScore(score);

    }
    /// <summary>
    /// Converts a MIDI note number to its corresponding name.
    /// </summary>
    /// <param name="noteNumber">The MIDI note number.</param>
    /// <returns>The note name (e.g., C, C#, D).</returns>
    public string ConvertNoteNumberToName(int noteNumber)
    {

        int octave = (noteNumber / 12) - 1;

        int noteIndex = (noteNumber) % 12;
        string noteName = noteNames[noteIndex];
        if (GameSettings.usePiano)
        {
            return $"{noteName} {octave}";
        }
        return $"{noteName}";
    }
    /// <summary>
    /// Converts a note name to its corresponding MIDI note number.
    /// </summary>
    /// <param name="noteName">The note name (e.g., C, C#, D).</param>
    /// <returns>The MIDI note number.</returns>
    public static int ConvertNoteNameToNumber(string noteName)
    {
        if (!char.IsDigit(noteName[0]))
        {
            int octave = 0;

            for (int i = 0; i < noteName.Length; i++)
            {
                if (char.IsDigit(noteName[i]))
                {
                    octave = noteName[i];
                }
            }
            nameToNoteMap.TryGetValue(noteName.ToLower(), out int value);

            return value + (12 * octave);
        }
        return int.Parse(noteName);

    }

    /// <summary>
    /// Coroutine for handling song end actions.
    /// </summary>
    public void OnSongEnd()
    {
        startTimer = false;
        MidiInput.instance.takeInput = false;
        int[] score = currentSongScore.GetScoreArray(totalNotes);

        if (!isCurSongPreview) { EndSongMessage.instance?.ShowScore($"Total Score: {score[0]}\nPerfect: {score[1]}\nGood: {score[2]}\nOkay: {score[3]}\nExtra: {score[4]}\nMissed: {score[5]}\nLongest Combo: {combo.highestCount}", currentSongScore.FinalizeScore()); }
        else { FindObjectOfType<EndPreview>()?.EndPreviewFade(); }

        if (!Replay.isPlayingReplay) { MidiDataHandler.SaveNoteEventData(".replay", Replay.instance.replayNoteData); } // only record replays if you arent playing back a replay.

    }



    /// <summary>
    /// Returns to the song selection scene.
    /// </summary>
    public void ReturnToSongSelection()
    {
        startTimer = false;
        inEditor = false;
        StopReadiedNotes();
        MidiInput.instance.inGame = false;
        GameSettings.ResetSettings(false);
        MidiInput.instance.UnHookMidiDevice();

        try { TransitionManager.instance.LoadNewScene("SongSelect"); }
        catch { SceneManager.LoadScene("SongSelect"); }

    }

    /// <summary>
    /// Stops all prepared notes and stops updating of songTime.
    /// </summary>
    public void StopSong()
    {
        startTimer = false;
        StopReadiedNotes();
        MidiInput.instance.inGame = false;
    }

   
    /*public void EnterSongEditor()
    {
        inEditor = true;
        string songName = GameSettings.currentSongName;
        if (string.IsNullOrEmpty(songName)) { return; }
        try { TransitionManager.instance.LoadNewScene("SongEditorScene"); }
        catch { SceneManager.LoadScene("SongEditorScene"); }
        MidiInput.instance.GetBPM(songName);
    }*/

    /// <summary>
    /// Reloads all json file versions of the currently selected song.
    /// </summary>
    public void RefreshJsonFiles()
    {
        NoteEventDataWrapper temp = MidiReadFile.GetNoteEventsFromFilePath(GameSettings.currentSongPath);
        MidiDataHandler.SaveNoteEventData(".json", temp.BPM, temp.NoteEvents);

    }

    /// <summary>
    /// Sets <see cref="beatsBeforeNote"/>.
    /// </summary>
    /// <param name="num">new <see cref="beatsBeforeNote"/>beatsBeforeNoteValue</param>
    public void SetBeatsBeforeDrop(string num)
    {
        if (!int.TryParse(num, out int newNum))
        {
            if (newNum <= 0) { Debug.LogWarning("BeatBeforeDrop Setting attempted to be set to <=0"); return; }
            Debug.LogError("Non Int Input into Beats Before Drop Setting"); return;
        }
        beatsBeforeNote = newNum;
        Debug.Log($"{newNum} Beats before note");
    }



}



